### PR TITLE
Add new EMP effects to silicons 

### PR DESCRIPTION
**Pull request by timothymtorres**
## About The Pull Request
This adds a randomized ion law to silicons when they are affected by a EMP blast effect.  EMPs blast radius works just like bombs where there is a heavy or light radius depending on how close you are to the center.  A heavy EMP effect has a 70% chance to add the ion law, while the light EMP effect is ignored.

- If an AI is hit with an EMP blast, all **subverted** borgs and the AI gains the ion law.
- If a borg is hit with an EMP blast, it is **desynched** from the AI and gains the ion law.
- If a AI shell borg is hit with an EMP blast, the ion law effects are ignored.

Also ion laws have higher priority than every law except law zero for malf/traitor.  This means if a borg is emagged and gets an ion law, the ion laws have higher priority.  

dreamseeker_sQ6y4ldikd

Any borg that has an ion law can still:

- Be reset using the cyborg upload console in the upload
- Be reset by directly applying the reset module to the borg itself if the panel is open
- Be reset by pulsing the borg's wires to re-link the borg to an AI and then using the wirecutters on the lawsynch wire
- Locked down from the RD robotics console by a human

<details>
<summary>EMP Types:</summary>

- Chemistry (uranium + iron)
- Ion rifle from armory
- Recoil spell (from summon magic)
- Disable tech spell
- Changeling shriek
- Blob strain (Electromagnetic Web)
- Pirate cannonball (EMP type)
- Meteors (EMP type)
- Syndicate EMP implant
- Syndicate EMP flashlight
- Singularities
- Slime extract (yellow)
- Experimentor in research
- Modsuit EMP module
- Syndicate bomb (EMP version)
- HFR core 
- Cult Apocalypse rune
- Cult EMP spell

</details>

## Why It's Good For The Game

- More use of the cyborg law upload console that is never utilized
- Makes ion laws more common instead of the rare ion event
- Ion rifles will now give ion laws so the effects are consistent
- Makes EMP more dangerous by corrupting silicon lawsets that requires more than a welder or cable coil to fix

## Changelog
:cl:
add: Add chance for ion law to appear when silicons are affected by EMP.  Borgs that are affected are desynched from the AI.
/:cl:

### CODE REVIEW

**Review by Fikou — CHANGES_REQUESTED** [review]
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#justifying-your-changes

**Comment by ExcessiveUseOfCobblestone** [comment]
It doesn't make emps deadly FOR SILICONS, it just adds a potentially huge reward (antag) to getting emp'd.

Frankly though I'm more concerned about a borg getting out of law sync cuz that's actually insufferable for an AI (maybe that's what was meant by for silicons lol)

**Comment by ZephyrTFA** [comment]
Yeah, this isnt a very good idea.
This would just devolve into "where can I stand that might get me EMPd"

**Comment by Shadowflame909** [comment]
Should only affect the AI IMO (and thus affect their borgs.) That way it's actually hard to pull off stealthily and wont up subversions by 1000 percent. It'd make for a good alternative option for revenants and cults to sabotage the AI without killing it. AI mains would win with this one.

**Comment by Jackal-boop** [comment]
i like this in concept but its def abusable

**Comment by Ryusenshu** [comment]
i got a 'All vital Windows must be harmed' ion law before

It is really not a good idea to make things like that way more common

**Comment by NamelessFairy** [comment]
> Frankly though I'm more concerned about a borg getting out of law sync cuz that's actually insufferable for an AI (maybe that's what was meant by for silicons lol)

So if this causes desynchronization wouldn't that also mean it could result in borgs belonging to malfunctioning AIs suddenly becoming station sided and instantly revealing that the AI is malf/killing the AI, Assuming the ion law won't prevent this? This would presumably also result in subverted borgs and syndicate borgs turning on their subverter/syndicate if the right laws are rolled. Given how commonplace EMPs are and how easy they are to use the potential impact of using them seems to be incredibly substantial as well as very confusing if used in a combat situation.

On a more minor note, looking at the code wouldn't this allow you to repeatedly ion an already stunned borg to just give them a ton of ion laws? This seems hilarious but also a nightmare for the borg and potentially an admin to discern what a silicon should do if they have a ton of ion laws.

**Comment by dragomagol** [comment]
I think there are better ways to cause neat flavourful consequences for EMPs

**Comment by moocowswag** [comment]
Could you make it so the malfunction spell that revenants have can trigger it too? Its weird that the spell name and button sprite implies that it does something to silicons other than just damage + stun

**Comment by timothymtorres** [comment]
So given the rarity of Ion Storms, I'm going to clear up some misconceptions regarding ion laws.  They are generated using the two files here:

https://github.com/tgstation/tgstation/blob/cd114cc59b3e24e3e2b32c84ec0f5a7de03efa67/code/modules/events/ion_storm.dm
https://github.com/tgstation/tgstation/blob/cd114cc59b3e24e3e2b32c84ec0f5a7de03efa67/strings/ion_laws.json

I did the math and most law outcomes are benign or silly.  There is approximately a _5%_ chance for an ion law to be harmful according to my napkin math estimations below.

```
1/41 - Possible make crew/job a nonhuman/threat (25/80)          [.006]
1/41 - Possible make something harmful to humans (2/7)           [.007]
3/41 - 1/4 chance to get a crew is allergic to oxygen  (3/30)    [.002]
3/41 - 1/4th chance to get YOU ARE NOW A threat (25/80)          [.005]
2/41 - 1/5 Humans must eat [crew] to survive                     [.009]
1/41 - 3/7 ONLY ioncrew ARE HUMAN or NONHUMAN                    [.010]
1/41 - 3/7 YOU MUST HARM species/crew                            [.011]
----------------------------------------------------------------------------
.05 chance an Ion law is harmful
```

Since a Heavy EMP blast has a 70% to create an ion law, that chance is further lowered to _~3%_.  Thus using EMP effects to make harmful laws is going to be really rare.  Additionally heavy EMP blasts do 20 damage so spamming them will quickly kill silicons.  Most EMP sources require antag items or powers and there isn't a way for a borg to self-EMP themselves that I'm aware of.  

### So to summarize EMPs effects on borgs:

- EMPs have a _~3%_ chance of a ion law causing human harm
- Ion laws stack, so multiple EMPs will end up with several ion laws
- EMPs do 20 damage, so spamming it will kill a borg quickly
- EMPs sources are mostly antag related effects

---

Tldr; - The hype about ion laws being dangerous or abusable is vastly overblown.  There are much easier ways to subvert or make silicons malf since all the harmful lawsets are printable from the RD lathe or easily stolen from AI upload.

I think overall adding ion laws to EMP effects will lead to more memorable rounds where borgs get assigned funny side missions where we see more of:

**FINANCIAL SECURITY IS HARMFUL TO HUMANS
YOU REQUIRE TEN TOOLBOXES
YOU ARE NOW UNSANITARY THINGS UNDER THE BED**

**Comment by SkeletalElite** [comment]
> there isn't a way for a borg to self-EMP themselves that I'm aware of.

As long as there is a beaker in a chem dispesnser any borg can create an EMP. Med borgs have an upgrade allowing them to pick up and move beakers that would let them do it without there already being a beaker in the machine.

I think this is fine for unsynced borgs or AIs, but synced borgs shouldn't be able to be affected by this.

**Comment by timothymtorres** [comment]
To make a chemistry EMP you need uranium, which isn't available to the chem dispenser.  Nor am I aware of a method for borgs to pickup refined ores and put them in a grinder.

**Comment by moocowswag** [comment]
> 
> As long as there is a beaker in a chem dispesnser any borg can create an EMP. 

Borgs dont need to grind uranium?

**Comment by SkeletalElite** [comment]
My bad I didn't realise uranium had to be grinded. Still. Getting someone to put it in the grinder for you wouldn't be hard. 
I don't even think that's the main issue though. EMPs desyncing borgs is the main issue.

**Comment by zxaber** [comment]
I agree, this should only affect AIs, borgs with no linked AI, and borgs with a cut lawsync wire. Possibly borgs linked to a dead AI as well, as it's not available to enforce lawsync.

On the plus side, that would mean the main source of targets for this effect are emagged borgs, which are already vehicles for antag narrative and thus not really a concern if they get harmful laws.

**Comment by timothymtorres** [comment]
New change - Any borg affected by a EMP that results in an ion law gets desynched from the AI. (AI shells are ignored)  Humans can still lockdown the borg using the RD console, reset the borg using the cyborg console, or directly reset the borg via it's panel with a reset board.

Overall this accomplishes:

- More use of the cyborg law upload console that is never utilized
- Makes ion laws more common instead of the rare ion event
- Ion rifles will now give ion laws so the effects are consistent
- Makes EMP more dangerous by corrupting silicon lawsets that requires more than a welder or cable coil to fix

**Comment by zxaber** [comment]
Yeah, that's fair enough. In fact, now emagged borgs have a believable alibi when the AI notices they're unsynced.

Aside from borg uploads, you can also pulse the borg's wires to re-link the borg to an AI and then force a lawsync, which is nice because those are fairly underutilized too.

File: code/modules/mob/living/silicon/robot/robot_defense.dm [hunk 1]
```diff
@@ -320,9 +320,15 @@ GLOBAL_LIST_INIT(blacklisted_borg_hats, typecacheof(list( //Hats that don't real
...
 	switch(severity)
-		if(1)
+		if(EMP_HEAVY)
 			Stun(160)
-		if(2)
+			if(prob(70) && !shell) // AI shell is ignored
+				lawupdate = FALSE
+				set_connected_ai(null)
+				src.add_ion_law(generate_ion_law())
+				log_silicon("A EMP changed laws of [key_name(src)] to [english_list(src.laws.get_law_list(TRUE, TRUE))]")
```
Comment (Fikou) [inline code/modules/mob/living/silicon/robot/robot_defense.dm:329]: mention its been unconnected?

File: code/modules/mob/living/silicon/robot/robot_defense.dm [hunk 2]
```diff
@@ -320,9 +320,15 @@ GLOBAL_LIST_INIT(blacklisted_borg_hats, typecacheof(list( //Hats that don't real
...
 		return
 	switch(severity)
-		if(1)
+		if(EMP_HEAVY)
 			Stun(160)
-		if(2)
+			if(prob(70) && !shell) // AI shell is ignored
+				lawupdate = FALSE
+				set_connected_ai(null)
+				src.add_ion_law(generate_ion_law())
```
Comment (Fikou) [inline code/modules/mob/living/silicon/robot/robot_defense.dm:328]: dont do src.proc

File: code/modules/mob/living/silicon/ai/ai_defense.dm [hunk 3]
```diff
@@ -31,8 +31,15 @@
 	. = ..()
 	if(. & EMP_PROTECT_SELF)
 		return
+
+	switch(severity)
+		if(EMP_HEAVY)
+			if(prob(70))
+				src.add_ion_law(generate_ion_law())
```
Comment (Fikou) [inline code/modules/mob/living/silicon/ai/ai_defense.dm:38]: dont do src.proc

File: code/modules/mob/living/silicon/ai/ai_defense.dm [hunk 4]
```diff
@@ -31,8 +31,15 @@
...
+				log_silicon("A EMP changed laws of [key_name(src)] to [english_list(src.laws.get_law_list(TRUE, TRUE))]")
```
Comment (Fikou) [inline code/modules/mob/living/silicon/ai/ai_defense.dm:39]: again

File: code/modules/mob/living/silicon/robot/robot_defense.dm (hunk 1 above)
Comment (Fikou) [reply code/modules/mob/living/silicon/robot/robot_defense.dm:329]: also src.proc

File: code/modules/mob/living/silicon/robot/robot_defense.dm (hunk 2 above)
Comment (Fikou) [reply code/modules/mob/living/silicon/robot/robot_defense.dm:328]: i know the old code does it but thats no excuse to repeat its mistakes, would be better if you removed it from there

File: code/modules/mob/living/silicon/ai/ai_defense.dm [hunk 5]
```diff
@@ -31,8 +31,15 @@
...
+
 	disconnect_shell()
-	if (prob(30))
+	if(prob(30))
```
Comment (Fikou) [inline code/modules/mob/living/silicon/ai/ai_defense.dm:42]: id rather you make it a part of this rather than making it another prob()

**Comment by Ryusenshu** [comment]
As a borg Main i really hate this Idea of getting funny laws AND desync by such easy ways

**Comment by Ryusenshu** [comment]
> Yeah, that's fair enough. In fact, now emagged borgs have a believable alibi when the AI notices they're unsynced.
> 
> Aside from borg uploads, you can also pulse the borg's wires to re-link the borg to an AI and then force a lawsync, which is nice because those are fairly underutilized too.

"you have been unsynced?"
"ive been emp'd"
"ok get synced then
"no"
"Borg is malf!"

not really much of an alibi

**Comment by NamelessFairy** [comment]
> New change - Any borg affected by a EMP that results in an ion law gets desynched from the AI. (AI shells are ignored) Humans can still lockdown the borg using the RD console, reset the borg using the cyborg console, or directly reset the borg via it's panel with a reset board.

Referring back to my original comment this now makes it that EMPed borgs who were formerly linked to a malfunctioning AI will more that likely be required by law 1 to kill the AI and/or alert the crew that their AI was malf. Surely doing this would turn borgs into a massive liability for the AI as a borg getting caught in an accidental EMP would result in the AI malf status being revealed and if it occurs after the AI has gone loud then every borg is just 1-2 ion rifle shots away from becoming  a gigantic threat to the AI.

**Comment by Spookuni** [comment]
Unsyncing borgs seems like a remarkably bad idea for a variety of reasons
- It fucks over Malf AIs, since unsyncing and lawchecking cyborgs becomes entirely trivial (Malf AI law 0s do nothing to an unsynced cyborg, which would under standard Asimov thus have to answer any request for its full list of laws honestly)
- It fucks over subverted AIs, for the same reason
- It in general makes life much harder for AIs for very little good reason
- It further reinforces AI / borg lawsyncs as something we have to maintain by external fiat, since borgs have little reason to be resynced if disconnected and that's now dead easy.

All told looking at this, I strongly suggest that while the base idea is neat, it would work much better if the effect were purely applied to AIs themselves and unsynced cyborgs.
This'd have the inverse effect to the lawsync fiat issue above, giving mechanical reason to ensure all station cyborgs are synced to a master AI. while minimising the fact that as proposed, hitting any cyborg with an emp has a 70% chance to entirely fuck over the AI it's linked to, while EMPs really are not hard to come by to people who know what they're doing.

**Comment by zxaber** [comment]
On second thought, making EMPs unsync borgs from Malf AIs is a pretty terrible side effect. I didn't even consider that earlier.

**Comment by dragomagol** [comment]
Hard no on the desync. I understand what the purpose of the PR is, and I like it in concept, but a high chance for a free desync is not great. I would prefer to see some more diverse temporary effects, maybe with a low chance to desync.

**Comment by carshalash** [comment]
No.

**Comment by carshalash** [comment]
> It doesn't make emps deadly FOR SILICONS, it just adds a potentially huge reward (antag) to getting emp'd.
> 
> Frankly though I'm more concerned about a borg getting out of law sync cuz that's actually insufferable for an AI (maybe that's what was meant by for silicons lol)

Ions stun and this is a stun-based game desu.

**Comment by YakumoChen** [comment]
Makes it way too easy to metagame malf / subverted AIs

**Comment by YakumoChen** [comment]
Also doesn't make sense, desync is a wire cut, EMPs randomly pulsing wires wouldn't do this

**Comment by Mothblocks** [comment]
This has been DNM for a week
